File: leetcode/0139-word-break/java/solution_2.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        return wordBreak(s, 0, wordDict, new HashMap<>());
    }

    private boolean wordBreak(String s, int index, List<String> wordDict, Map<Integer, Boolean> memo) {
        
        if (index == s.length()) {
            return true;
        }

        if (memo.containsKey(index)) {
            return memo.get(index);
        }

        boolean res = false;
        for (String word : wordDict) {
            if (s.startsWith(word, index)) {
                res |= wordBreak(s, index + word.length(), wordDict, memo);
            }
        }

        memo.put(index, res);
        return res;
    }
}

File: leetcode/0139-word-break/rust/solution_2.rs
use std::collections::HashMap;

impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let mut memo: HashMap<usize, bool> = HashMap::new();
        Solution::word_break_search(&s, 0, &word_dict, &mut memo)
    }

    fn word_break_search(s: &str, index: usize, word_dict: &Vec<String>, memo: &mut HashMap<usize,bool>) -> bool {
        if index == s.len() {
            return true;
        }

        if let Some(&cached) = memo.get(&index) {
            return cached;
        }

        for word in word_dict {
            if s[index..].starts_with(word) {
                if Solution::word_break_search(s, index + word.len(), word_dict, memo) {
                    return true;
                }
            }
        }

        memo.insert(index, false);
        false
    }
}